Undo and redo command

KellyLi opened the pull request.

KellyLi committed:
Adding undo command for add, delete, and edit

KellyLi committed:
Add redo command

KellyLi committed:
Tests for undo

KellyLi committed:
Adding redo tests

KellyLi committed:
Adding tests for redo

KellyLi committed:
Add undo test

evanyeung commented on src/main/java/seedu/task/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,53 @@
+        final Optional<TaskMemento> memento = mementos.getRedoMemento();
+        final Task mementoTask = memento.orElseThrow(
+            () -> new CommandException(MESSAGE_NO_HISTORY)).newTask;
+        final TaskId mementoTaskId = memento.orElse(null).taskId;
```
This is going to cause a null pointer error if memento is empty. It kinda defeats the point of the optional...

evanyeung commented on src/main/java/seedu/task/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,53 @@
+    public CommandResult execute() throws CommandException {
+        final Optional<TaskMemento> memento = mementos.getRedoMemento();
+        final Task mementoTask = memento.orElseThrow(
+            () -> new CommandException(MESSAGE_NO_HISTORY)).newTask;
```
Fancy O_o

evanyeung commented on src/main/java/seedu/task/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,53 @@
+            try {
+                model.deleteTask(taskToBeReplaced);
+            } catch (TaskNotFoundException e) {
+                assert false : "The target task cannot be missing";
```
Is there a point of this assert? The error message should just be in the `TaskNotFoundException` which, if your not handling it, the function can throw.

KellyLi replied:
It should technically be an assert because you should never be deleting a task that was never added

evanyeung replied:
Okay, I see how these are different from throwing command errors

evanyeung commented on src/main/java/seedu/task/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,53 @@
+            try {
+                model.addTask(mementoTask);
+            } catch (DuplicateTaskException e) {
+                assert false : "Adding duplicate task";
```
^^

evanyeung commented on src/main/java/seedu/task/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,53 @@
+            try {
+                model.updateTaskById(mementoTaskId, mementoTask);
+            } catch (DuplicateTaskException e) {
+                assert false : "duplicate task";
```
![^^](http://assets.pokemon.com/assets/cms2/img/pokedex/full/132.png)

evanyeung commented on src/main/java/seedu/task/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,53 @@
+    public static final String MESSAGE_NO_HISTORY = "No commands to redo";
+
+    @Override
+    public CommandResult execute() throws CommandException {
```
Should you assert that both commands are not null? You don't check for that case in the function body and a successful `CommandResult` would be returned.

KellyLi replied:
Sorry which commands?

evanyeung replied:
Oh, I meant that both tasks within the memento are not null. Sorry, don't know why I put commands

evanyeung commented on src/main/java/seedu/task/logic/commands/UndoCommand.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Undo previous commands executed
+ */
+public class UndoCommand extends Command {
```
Same comments as on `RedoCommand`

evanyeung commented on src/main/java/seedu/task/logic/history/TaskMemento.java:
```diff
@@ -14,23 +14,22 @@
-        this.oldTask = oldTask;
-        this.newTask = newTask;
+        this.oldTask = oldTask != null ? new Task(oldTask) : null;
+        this.newTask = newTask != null ? new Task(newTask) : null;
```
Thanks :]

evanyeung commented on src/main/java/seedu/task/logic/history/TaskMementos.java:
```diff
@@ -23,7 +23,7 @@ public TaskMementos() {
      */
     public void addUndoMementoAndClearRedo(TaskMemento taskMemento) {
-        this.undoMementoStack.push(taskMemento);
+        this.undoMementoStack.push(new TaskMemento(taskMemento.oldTask, taskMemento.newTask));
```
Do you have to copy here? The `TaskMemento` should be immutable anyway and the tasks are copied in the `TaskMemento` constructor.

KellyLi replied:
Yeah I don't think it needs to be copied, ty!

evanyeung reviewed (changes requested):
Looks good but commented on a few small things

KellyLi committed:
Code review

KellyLi committed:
Code review

- Assert if both tasks in memento are null for undo and redo
- Throw error for null memento returned

evanyeung reviewed (approved):
![](https://media.giphy.com/media/yyZRSvISN1vvW/giphy.gif)
